csrc/multidevice/communication.cpp

```diff
@@ -378,31 +378,47 @@ c10::intrusive_ptr<c10d::Work> postScatter(
     c10d::Backend* backend,
     at::Tensor input_tensor,
     at::Tensor output_tensor) {
-  if (my_device_index == communication->root() &&
-      !communication->out()->getDeviceMesh().has(communication->root())) {
-    output_tensor = at::empty_like(input_tensor.slice(0, 0, 1));
-  }
-  std::vector<at::Tensor> output_tensors({output_tensor});
+  NVF_ERROR(
+      isTvContiguous(communication->in()), "Input tensor is not contiguous");
+  NVF_ERROR(
+      isTvContiguous(communication->out()), "Output tensor is not contiguous");
 
+  auto output_device_mesh = communication->out()->getDeviceMesh();
+  bool output_has_root = output_device_mesh.has(communication->root());
   auto root_relative_index = communication->getRootRelativeIndex();
+
   std::vector<std::vector<at::Tensor>> input_tensors;
+
+  if (output_tensor.defined()) {
+    output_tensor = output_tensor.as_strided({output_tensor.numel()}, {1});
+  }
+
   if (my_device_index == communication->root()) {
+    auto splits = at::tensor_split(
+        input_tensor.as_strided({input_tensor.numel()}, {1}),
+        output_device_mesh.size(),
+        /*dim=*/0);
+    if (!output_has_root) {
+      output_tensor = at::empty_like(splits.at(0));
+    }
     input_tensors.resize(1);
     int64_t j = 0;
     for (auto i : arange(communication->team().size())) {
       if (root_relative_index == static_cast<DeviceIdxType>(i) &&
-          !communication->out()->getDeviceMesh().has(communication->root())) {
-        input_tensors.front().push_back(output_tensor);
+          !output_has_root) {
+        input_tensors.front().push_back(at::empty_like(splits.at(0)));
```

Comment:
Is there a shape mismatch if you keep using output_tensor?

Reply:
No, it should be fine.
Personally, it was easier for me to follow the code in terms of `splits`. Hence, the change. I can switch it back.

Reply:
Got it. I asked because this line creates an extra tensor in addition to the one created on line 402. 

(Again, if you or someone else is going to unsupport !output_has_root, the change doesn't matter)

Reply:
Yes, I am removing that condition in the latest changes. Running tests now to verify correctness.